presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -851,7 +827,7 @@ public PlanWithProperties visitSemiJoin(SemiJoinNode node, Context context)
             PlanWithProperties source;
             PlanWithProperties filteringSource;
 
-            if (distributedJoins && !context.isDownstreamIsDelete()) {
+            if (node.getDistributionType().orElseThrow(() -> new IllegalStateException("distributionType not yet set")) == SemiJoinNode.DistributionType.PARTITIONED) {
```

Comment:
Same as for the other commit ... move the illegal state check outside of the if condition.